### PR TITLE
Add document for CTE

### LINKED ISSUE #6216: Add document for CTE
## Change Request

Add document for CTE, this issue is used for tracing.

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

Close #6216 
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Comment by wjhuang2016** [comment]
This PR is not ready to review now.

**Comment by TomShawn** [comment]
@wjhuang2016 Please involve a technical review(s). Thanks~

File: sql-statements/sql-statement-with.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+---
+title: WITH (公共表表达式)
+summary: TiDB 数据库中 WITH (公共表表达式) 的使用概况。
+---
+
+# WITH (公共表表达式)
+
+`公共表表达式` 是一个临时的中间结果集，能够在语句中多次引用。
```
Comment (qiancai) [inline sql-statements/sql-statement-with.md:8]:
```suggestion
公共表表达式 (CTE) 是一个临时的中间结果集，能够在 SQL 语句中引用多次，提高 SQL 语句的可读性与执行效率。
```

File: sql-statements/sql-statement-with.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+| 3 |
+| 4 |
+| 5 |
++---+
+5 rows in set (0.00 sec)
+```
+
+## MySQL 兼容性
+
+* 在严格模式下，当递归部分算出的数据的长度超过初始部分的长度时，TiDB 会返回警告, 而 MySQL 会返回错误。非严格模式下，TiDB 与 MySQL 一致。
```
Comment (qiancai) [inline sql-statements/sql-statement-with.md:70]:
```suggestion
* 在严格模式下，当递归部分算出的数据长度超过初始部分的数据长度时，TiDB 会返回警告，而 MySQL 会返回错误。在非严格模式下，TiDB 与 MySQL 行为一致。
```

File: sql-statements/sql-statement-with.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+* 递归 CTE 所使用的数据类型由初始部分决定。初始部分的类型在某些情况（例如函数）下并不与 MySQL 完全一致。
+* 多个 UNION / UNION ALL 情况下，MySQL 不允许 UNION 后面加 UNION ALL ，TiDB 允许。
```
Comment (qiancai) [inline sql-statements/sql-statement-with.md:72]:
```suggestion
* 多个 UNION / UNION ALL 情况下，MySQL 不允许 UNION 后面加 UNION ALL，TiDB 允许。
```

**Review by xiongjiwei — APPROVED** [review]

File: sql-statements/sql-statement-with.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+![CommonTableExpr](/media/sqlgram/CommonTableExpr.png)
+
+**CommonTableExpr:**
+
+![CommonTableExpr](/media/sqlgram/CommonTableExpr.png)
+
+**IdentListWithParenOpt:**
+
+![IdentListWithParenOpt](/media/sqlgram/IdentListWithParenOpt.png)
```
Comment (TomShawn) [inline sql-statements/sql-statement-with.md:26]: 这里建议用 ebnf 语法，官网会自动渲染出图

```ebnf
```

File: sql-statements/sql-statement-with.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
+---
+title: WITH (公共表表达式)
+summary: TiDB 数据库中 WITH (公共表表达式) 的使用概况。
+---
+
+# WITH (公共表表达式)
```
Comment (TomShawn) [inline sql-statements/sql-statement-with.md:6]:
```suggestion
# WITH
```

File: sql-statements/sql-statement-with.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+公共表表达式 (CTE) 是一个临时的中间结果集，能够在 SQL 语句中引用多次，提高 SQL 语句的可读性与执行效率。
```
Comment (TomShawn) [inline sql-statements/sql-statement-with.md:8]:
```suggestion
公共表表达式 (CTE) 是一个临时的中间结果集，能够在 SQL 语句中引用多次，提高 SQL 语句的可读性与执行效率。在 TiDB 中可以通过 `WITH` 语句使用公共表达式。
```

Comment (wjhuang2016) [reply sql-statements/sql-statement-with.md:8]: 是公共表表达式，而不是公共表达式

**Review by TomShawn — APPROVED** [review]
LGTM

File: sql-statements/sql-statement-with.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+| 4 |
+| 5 |
++---+
+5 rows in set (0.00 sec)
+```
+
+## MySQL 兼容性
+
+* 在严格模式下，当递归部分算出的数据长度超过初始部分的数据长度时，TiDB 会返回警告，而 MySQL 会返回错误。在非严格模式下，TiDB 与 MySQL 行为一致。
+* 递归 CTE 所使用的数据类型由初始部分决定。初始部分的类型在某些情况（例如函数）下并不与 MySQL 完全一致。
```
Comment (qiancai) [inline sql-statements/sql-statement-with.md:86]:
```suggestion
* 递归 CTE 所使用的数据类型由初始部分决定。初始部分的数据类型在某些情况（例如函数）下与 MySQL 并不完全一致。
```

Comment (qiancai) [inline sql-statements/sql-statement-with.md:86]: 关于 “初始部分的类型在某些情况（例如函数）下并不与 MySQL 完全一致。”，这里请问可以再具体说下吗“某些情况”是哪些情况吗？只有函数是我们已知的不一致的情况吗？这种不一致是否会对计算结果产生影响呢？

Comment (wjhuang2016) [reply sql-statements/sql-statement-with.md:86]: 这些情况并不能列举完，会对计算结果有影响。

Comment (qiancai) [reply sql-statements/sql-statement-with.md:86]: 不一致的情况可以再提供一个例子吗？当无法直接说明是哪些情况时候，能够多给几个例子对用户的帮助也是极大的。

Comment (wjhuang2016) [reply sql-statements/sql-statement-with.md:86]: 暂时没想到

**Review by qiancai — APPROVED** [review]

**Pull request merged**